aten/src/ATen/native/Loss.cpp

```diff
@@ -52,23 +51,38 @@ Tensor triplet_margin_loss(const Tensor& anchor, const Tensor& positive, const T
     dist_neg = at::min(dist_neg, dist_swap);
   }
   auto output = at::clamp_min(margin + dist_pos - dist_neg, 0);
-
-  if (reduction == Reduction::ElementwiseMean) {
-    return output.sum() / output.numel();
-  } else if (reduction == Reduction::Sum) {
-    return output.sum();
-  }
-  return output;
+  return apply_loss_reduction(output, reduction);
 }
 
 Tensor margin_ranking_loss(const Tensor& input1, const Tensor& input2, const Tensor& target, double margin, int64_t reduction) {
   auto output =  (-target * (input1 - input2) + margin).clamp_min_(0);
+  return apply_loss_reduction(output, reduction);
+}
 
+Tensor kl_div(const Tensor& input, const Tensor& target, int64_t reduction) {
+  auto zeros = at::zeros_like(target);
+  auto output_pos = target * (at::log(target) - input);
+  auto output = at::where(target > 0, output_pos, zeros);
+  return apply_loss_reduction(output, reduction);
+}
+
+Tensor kl_div_backward_cpu(const Tensor& grad, const Tensor& input, const Tensor& target, int64_t reduction) {
+  auto grad_input = grad.type().zeros_like(input);
+  auto grad_expand = grad.expand_as(input);
+  AT_DISPATCH_FLOATING_TYPES(input.type(), "kl_div_backward", [&]() {
+    at::CPU_tensor_apply3<scalar_t, scalar_t, scalar_t>(
+        grad_input,
+        target,
+        grad_expand,
+        [] (scalar_t& grad_input_val, const scalar_t& target_val, const scalar_t& grad_val) {
+          if (target_val > 0) {
+            grad_input_val = -target_val * grad_val;
+          }
+        });
+  });
   if (reduction == Reduction::ElementwiseMean) {
-    return output.sum() / output.numel();
-  } else if (reduction == Reduction::Sum) {
-    return output.sum();
+    return grad_input / input.numel();
```

Comment:
The case for `Reduction::Sum` seems removed. I didn't check the maths, but am wondering if the code works for both `Sum` and `None` cases.

Reply:
Yeah, that's intended. We only need to divide by numel in this case. The expand conveniently allows us to deal with the other two cases simultaneously.